File: sdk/approute-sdk/javascript/src/resources/funds.ts
import { BaseResource } from "./base-resource.js";
import type {
  BybitState,
  FundingInvoice,
  FundingInvoiceList,
  FundingInvoiceTimeLeft,
  FundingMethodsResponse,
  TonDepositState,
} from "../models/index.js";

export interface CreateInvoiceOptions {
  methodCode: string;
  amount: number;
}

export interface ListInvoicesOptions {
  status?: string[];
  methodCode?: string[];
  search?: string;
  invoiceId?: string;
  createdFrom?: string;
  createdTo?: string;
  withTx?: boolean;
  limit?: number;
  offset?: number;
}

/**
 * Resource for the /funds endpoints (funding methods and invoices).
 */
export class FundsResource extends BaseResource {
  /**
   * List available funding methods.
   */
  async methods(): Promise<FundingMethodsResponse> {
    return this.transport.request<FundingMethodsResponse>(
      "GET",
      "/funds/methods",
    );
  }

  /**
   * Create a funding invoice.
   */
  async createInvoice(options: CreateInvoiceOptions): Promise<FundingInvoice> {
    return this.transport.request<FundingInvoice>("POST", "/funds/invoices", {
      body: {
        methodCode: options.methodCode,
        amount: options.amount,
      },
    });
  }

  /**
   * List funding invoices with filtering.
   */
  async listInvoices(
    options?: ListInvoicesOptions,
  ): Promise<FundingInvoiceList> {
    const params: Record<string, unknown> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };

    if (options?.status !== undefined) params.status = options.status;
    if (options?.methodCode !== undefined) params.methodCode = options.methodCode;
    if (options?.search !== undefined) params.search = options.search;
    if (options?.invoiceId !== undefined) params.invoiceId = options.invoiceId;
    if (options?.createdFrom !== undefined) params.createdFrom = options.createdFrom;
    if (options?.createdTo !== undefined) params.createdTo = options.createdTo;
    if (options?.withTx !== undefined) params.withTx = options.withTx;

    return this.transport.request<FundingInvoiceList>(
      "GET",
      "/funds/invoices",
      { params },
    );
  }

  /**
   * Get a single funding invoice by ID.
   */
  async getInvoice(invoiceId: string): Promise<FundingInvoice> {
    return this.transport.request<FundingInvoice>(
      "GET",
      `/funds/invoices/${invoiceId}`,
    );
  }

  /**
   * Check (refresh) the status of a funding invoice.
   */
  async checkInvoice(invoiceId: string): Promise<FundingInvoice> {
    return this.transport.request<FundingInvoice>(
      "POST",
      `/funds/invoices/${invoiceId}/check`,
    );
  }

  /**
   * Get time left before a funding invoice expires.
   */
  async invoiceTimeLeft(invoiceId: string): Promise<FundingInvoiceTimeLeft> {
    return this.transport.request<FundingInvoiceTimeLeft>(
      "GET",
      `/funds/invoices/${invoiceId}/time-left`,
    );
  }

  /**
   * Get TON deposit address and memo tag.
   */
  async tonDeposit(): Promise<TonDepositState> {
    return this.transport.request<TonDepositState>(
      "GET",
      "/funds/ton/deposit",
    );
  }

  /**
   * Get Bybit UID state.
   */
  async bybitState(): Promise<BybitState> {
    return this.transport.request<BybitState>("GET", "/funds/bybit/state");
  }

  /**
   * Attach a Bybit UID.
   */
  async bybitAttach(uid: string): Promise<BybitState> {
    return this.transport.request<BybitState>("POST", "/funds/bybit/attach", {
      body: { uid },
    });
  }

  /**
   * Unlink the attached Bybit UID.
   */
  async bybitUnlink(): Promise<BybitState> {
    return this.transport.request<BybitState>("POST", "/funds/bybit/unlink");
  }
}

File: sdk/approute-sdk/python/approute/resources/funds.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from approute.models import (
    BybitState,
    FundingInvoice,
    FundingInvoiceList,
    FundingInvoiceTimeLeft,
    FundingMethodsResponse,
    TonDepositState,
)
from approute.resources._base import BaseResource


class FundsResource(BaseResource):

    def methods(self) -> FundingMethodsResponse:
        """List available funding methods."""
        data = self._t.request("GET", "/funds/methods")
        return FundingMethodsResponse.model_validate(data)

    def create_invoice(self, *, method_code: str, amount: float) -> FundingInvoice:
        """Create a funding invoice."""
        data = self._t.request(
            "POST",
            "/funds/invoices",
            json_body={"method_code": method_code, "amount": amount},
        )
        return FundingInvoice.model_validate(data)

    def list_invoices(
        self,
        *,
        status: list[str] | None = None,
        method_code: list[str] | None = None,
        search: str | None = None,
        invoice_id: str | None = None,
        created_from: datetime | None = None,
        created_to: datetime | None = None,
        with_tx: bool | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> FundingInvoiceList:
        """List funding invoices."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status
        if method_code is not None:
            params["method_code"] = method_code
        if search is not None:
            params["search"] = search
        if invoice_id is not None:
            params["invoice_id"] = invoice_id
        if created_from is not None:
            params["created_from"] = created_from.isoformat()
        if created_to is not None:
            params["created_to"] = created_to.isoformat()
        if with_tx is not None:
            params["with_tx"] = with_tx
        data = self._t.request("GET", "/funds/invoices", params=params)
        return FundingInvoiceList.model_validate(data)

    def get_invoice(self, invoice_id: str) -> FundingInvoice:
        """Get a single funding invoice by ID."""
        data = self._t.request("GET", f"/funds/invoices/{invoice_id}")
        return FundingInvoice.model_validate(data)

    def check_invoice(self, invoice_id: str) -> FundingInvoice:
        """Check (refresh) the status of a funding invoice."""
        data = self._t.request("POST", f"/funds/invoices/{invoice_id}/check")
        return FundingInvoice.model_validate(data)

    def invoice_time_left(self, invoice_id: str) -> FundingInvoiceTimeLeft:
        """Get time left before a funding invoice expires."""
        data = self._t.request("GET", f"/funds/invoices/{invoice_id}/time-left")
        return FundingInvoiceTimeLeft.model_validate(data)

    def ton_deposit(self) -> TonDepositState:
        """Get TON deposit address and memo tag."""
        data = self._t.request("GET", "/funds/ton/deposit")
        return TonDepositState.model_validate(data)

    def bybit_state(self) -> BybitState:
        """Get Bybit UID state."""
        data = self._t.request("GET", "/funds/bybit/state")
        return BybitState.model_validate(data)

    def bybit_attach(self, uid: str) -> BybitState:
        """Attach a Bybit UID."""
        data = self._t.request("POST", "/funds/bybit/attach", json_body={"uid": uid})
        return BybitState.model_validate(data)

    def bybit_unlink(self) -> BybitState:
        """Unlink the attached Bybit UID."""
        data = self._t.request("POST", "/funds/bybit/unlink")
        return BybitState.model_validate(data)
